### ISSUE TITLE
False positive match in Glossary term '&'

### ISSUE THREAD

**Issue by pedro-mendonca** [GlotPress/GlotPress-WP#1765]
## Description
In the test `test_map_glossary_entries_to_translation_originals_with_ampersand_in_glossary()` there is a glossary match that seems to be a false positive.

It was discussed first here https://github.com/GlotPress/GlotPress/issues/869 but I think it needs a second look.
The string `This string, <code>&lt;/body&gt;</code>, should not have the code tags mangled.` on the test happens to match the glossary term '&' only on the second appearance, which is wrong.
On the frontend this match doesn't happen.

To start, it shouldn't ever match `&` if the string is `&lt;` or `&gt;`, only on a string like `Products & Services`.

But for that, there is an issue with the current regex to match terms: it's something like `\b(term_a|term_b|etc)\b`, which would easily catch it.... but `&` isn't a word character, so even on this situation, won't ever be matched.

Another issue: It's not currently possible to add `&` to the glossary, because of the same reason: the string is not starting or ending, the actual character isn't.... a word character.

What would be the best solution for this?

**Comment by amieiro** [comment]
Currently, it is not possible to create a new record in a glossary with the "&" word. You get this error: The field term is invalid and should be starting and ending with a word character!

![image](https://github.com/GlotPress/GlotPress/assets/1667814/49cb905e-be6e-47f7-8c9e-76fb37800c6e)

![image](https://github.com/GlotPress/GlotPress/assets/1667814/d16d7528-e0fb-40b9-a703-a7151dab65df)

I have created it, updating a previous record in the database with the "&" symbol.

![image](https://github.com/GlotPress/GlotPress/assets/1667814/3957782d-ce23-405d-a20b-0e4ae89b63df)

Currently, it doesn't detect the "&" element inside other words, neither as unique words.

![image](https://github.com/GlotPress/GlotPress/assets/1667814/7780e4c9-f925-42dd-afbc-f5d1b8a29f74)

The [test_map_glossary_entries_to_translation_originals_with_ampersand_in_glossary](https://github.com/GlotPress/GlotPress/blob/3e0ddb23f22360233877c650670afad4c9c8ec7f/tests/phpunit/testcases/test_template_helper_functions.php#L8-L30) test is incorrect, because detects one of the "&" items for the string 

```
This string, <code>&lt;/body&gt;</code>, should not have the code tags mangled.
```

As no one of the "&" characters make a whole word, I think this test should not detect any "&" inside the glossary.

But I think this is an edge case and it has very little use as a complete word, so I think we should remove support for it. If we consider supporting it, we should consider [English terms with diacritical marks](https://en.wikipedia.org/wiki/English_terms_with_diacritical_marks) : words like `piñata` or `jalapeño` from Spanish, `frappé` from French, ...

**Comment by akirk** [comment]
I understand the only use case for an ampersand would be one that's surrounded with whitespace. Would it work to rewrite it to ` & ` and/or `&amp;` before we detect it in a string as a glossary term?

**Comment by amieiro** [comment]
> I understand the only use case for an ampersand would be one that's surrounded with whitespace. Would it work to rewrite it to `&` and/or `&amp;` before we detect it in a string as a glossary term?

Yes, this is the unique possibility. @pedro-mendonca is working to improve it in the https://github.com/GlotPress/GlotPress/pull/1766 PR, so let's see if we can get this improvement, to detect only these elements surrounded with whitespace.

**Comment by pedro-mendonca** [comment]
I just started to investigate and collect all the possible cases. For now it's up to be decided wether or not to allow `&` (or other symbols), as any development for it will need to add some specific exception, because actually `&` isn't a word character with `\b` (word bounds). 

More or less the same needs to be decided about diacritcs.
Should it be possible to add non-english borrowed words as glossary entries? (some examples added to tests in https://github.com/GlotPress/GlotPress/pull/1766/commits/ecab15484d2ef19ad253e53f1fe841d244251c8a) I see it can be useful, still, I would like to have your opinion on this before start digging for a solution.

**Comment by ekamran** [comment]
I checked how the boundary change in #2087 (same lookarounds as draft #1766) behaves against this issue. The named-entity behaviour is asserted by tests in that PR; the numeric-entity results below are from a local test harness run against the same branch.

Fixed: the reported case. The `&` term no longer matches inside named references like `&lt;`/`&gt;`, because the character after `&` is a letter, which the lookahead rejects. A standalone `&` still matches, asserted in the PR with `Categories & Tags` and verified locally with `Products & Services` as well.

Also fixed: `&` can now be added to a glossary. The `starting_and_ending_with_a_word_character` rule that blocked it, reported here and confirmed by amieiro, is removed in that PR and covered by tests.

Still false-positive: numeric character references, in all three forms.

- `&#38;` - the `&` matches
- `&#8212;` - the `&` matches
- `&#x26;` - the `&` matches

After `&` comes `#`, which is not a word character, so the lookahead passes. Fixing that needs entity-aware matching rather than boundary tuning, since a generic exclusion of `#` would also change matching for unrelated terms. So #2087 narrows this issue but does not close it.
